### PR TITLE
refactor(submenu mode vertical or horizontal added): submenu mode

### PR DESCRIPTION
**Pull request by govindsingh55**
submenu mode vertical or horizontal added

### CODE REVIEW

**Comment by navin-moorthy** [comment]
@govindsingh55, I saw your implementation. Progressing good, the mode prop should be for MenuBar, not the SubMenu.

image


Now the SubMenu is Vertical and outside PopOver, but the MenuBarItems are still horizontal.

We need the MenuBarItems as-well in Vertical like this,

image

**Comment by govindsingh55** [comment]
> @govindsingh55, I saw your implementation. Progressing good, the mode prop should be for MenuBar, not the SubMenu.

add the flexDirection="column" to the MenuBar

**Comment by navin-moorthy** [comment]
Adding flexDirection="column" to the MenuBar should be done in the component based on the mode prop.

**Comment by govindsingh55** [comment]
but will it contradict when the menubar is vertical one but we want the submenu to open in popper?

**Comment by navin-moorthy** [comment]
No that will not contradict that, default behavior on mode=`vertical` makes menubar flexdirection column and the submenu still remains popper but the Popper `placement`: `right`.
Also, when `isCollapsable` only on mode=`vertical`, then instead of popper another component for the collapsed item will be rendered.
Understood?

**Comment by govindsingh55** [comment]
I am not considering the navbar or hamburger case.
In the case of side menubar, submenu can be a popover or collapsable container right?

**Comment by navin-moorthy** [comment]
yeah @govindsingh55 , even I am not considering the navbar.
Yes. in case of side menubar, submenu can be popover by default but on `isCollapsable` it becomes a collapsable container.
That's correct.

**Comment by govindsingh55** [comment]
so isCollapsable prop will decide to popper or collapsable container.

**Comment by navin-moorthy** [comment]
> so isCollapsable prop will decide to popper or collapsable container.

Correct 💯

File: packages/chakra-components/src/MenuBar/SubMenu.js [hunk 1]
```diff
@@ -2,7 +2,6 @@ import React, { createContext, useState, useRef, useEffect } from "react";
 import { PseudoBox, usePrevious, useColorMode } from "@chakra-ui/core";
 import { useId } from "@reach/auto-id";
 
-import { useMenuBarContext } from "./useMenuBarContext";
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/SubMenu.js:5]: Why was this removed?

File: packages/chakra-components/src/MenuBar/SubMenu.js [hunk 2]
```diff
@@ -34,7 +34,6 @@ const SubMenu = ({
 
   const _isOpen = isControlled ? isOpenProp : isOpen;
   const wasPreviouslyOpen = usePrevious(_isOpen);
-  const wasPreviouslyOpenBeforeTimeout = usePrevious(wasPreviouslyOpen);
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/SubMenu.js]: Why was this removed?

File: packages/chakra-components/src/MenuBar/SubMenu.js [hunk 3]
```diff
@@ -45,8 +44,6 @@ const SubMenu = ({
 
   const { colorMode } = useColorMode();
 
-  const { trigger } = useMenuBarContext();
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/SubMenu.js:48]: Why was this removed?

File: packages/chakra-components/src/MenuBar/SubMenu.js [hunk 4]
```diff
@@ -67,34 +64,14 @@ const SubMenu = ({
...
-    }
-
-    if (trigger === "hover") {
-      if (
-        activeIndex === -1 &&
-        !_isOpen &&
-        (wasPreviouslyOpenBeforeTimeout || wasPreviouslyOpen)
-      ) {
-        titleRef.current && titleRef.current.focus();
-      }
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/SubMenu.js]: Why was this removed?

File: packages/chakra-components/src/MenuBar/SubMenu.js [hunk 5]
```diff
@@ -67,34 +64,14 @@ const SubMenu = ({
...
-  }, [
-    activeIndex,
-    _isOpen,
-    titleRef,
-    menuRef,
-    trigger,
-    wasPreviouslyOpen,
-    wasPreviouslyOpenBeforeTimeout,
-  ]);
+  }, [activeIndex, _isOpen, titleRef, menuRef, wasPreviouslyOpen]);
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/SubMenu.js]: Why was this removed?

File: packages/chakra-components/src/MenuBar/SubMenu.js [hunk 6]
```diff
@@ -159,26 +136,13 @@ const SubMenu = ({
...
-      setIsOpen(false);
-    }
-
-    if (onClose) {
-      onClose();
-    }
-
-    resetTabIndex();
-  };
-
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/SubMenu.js]: Why was this removed?

File: packages/chakra-components/src/MenuBar/SubMenu.js [hunk 7]
```diff
@@ -159,26 +136,13 @@ const SubMenu = ({
...
   const context = {
     activeIndex,
     isOpen: _isOpen,
     focusAtIndex,
     focusOnLastItem,
     focusOnFirstItem,
     closeMenu,
-    closeMenuWithoutIndex,
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/SubMenu.js:181]: Why was this removed?

File: packages/chakra-components/src/MenuBar/SubMenu.js [hunk 8]
```diff
@@ -190,13 +154,24 @@ const SubMenu = ({
...
   };
 
   return (
     <SubMenuContext.Provider value={context}>
-      <PseudoBox as="li" role="none" display="flex" {...props}>
+      <PseudoBox
+        as="li"
+        role="none"
+        display="flex"
+        flexDirection={isCollapsable ? "column" : "row"}
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/SubMenu.js]: Instead of creating the style prop like this.

Create a separate condition about return like below

```suggestion
let collapsableStyleProps = {}
if(isCollapsable) {
collapsableStyleProps = {flexDirection: "column" }}
}
```
 In Pseudobox after {...styleProps} include this as {...collapsableStyleProps}.

We are doing this so the base styles are overridden properly

File: packages/chakra-components/src/MenuBar/SubMenu.js [hunk 9]
```diff
@@ -190,13 +154,24 @@ const SubMenu = ({
...
+        {...props}
+      >
         {typeof children === "function"
-          ? children({ isOpen: _isOpen, onClose: closeMenu })
+          ? children({
+              isOpen: _isOpen,
+              onClose: closeMenu,
+              isCollapsable: isCollapsable,
+            })
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/SubMenu.js]: We dont want to allow, isCollapsable in renderProps. IMO this is unnecessary since we are already controlling it via prop

File: packages/chakra-components/src/MenuBar/SubMenuList.js [hunk 10]
```diff
@@ -151,35 +152,48 @@ const SubMenuList = ({
...
+        as="ul"
+        ref={menuRef}
+        width={width}
+        display={isOpen ? "block" : "none"}
+        onKeyDown={handleKeyDown}
+        {...eventHandlers}
+        {...props}
+      />
+    );
+  }
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/SubMenuList.js]:
```suggestion
  if (isCollapsable) {
    return (
      <Box
        as="ul"
        ref={menuRef}
        width={width}
        display={isOpen ? "block" : "none"}
        onKeyDown={handleKeyDown}
        {...eventHandlers}
        {...props}
      />
    );
  }
    return (
      <Popper
        usePortal={false}
        as={Comp}
        anchorEl={titleRef.current}
        ref={menuRef}
        isOpen={isOpen}
        placement={placement}
        modifiers={popperModifiers}
        width={width}
        rounded="md"
        py={2}
        zIndex="2"
        _focus={{
          outline: 0,
        }}
        role="menu"
        aria-label={ariaLabel}
        tabIndex={-1}
        onKeyDown={handleKeyDown}
        onBlur={handleBlur}
        {...eventHandlers}
        {...styleProps}
        {...props}
      />
    );
```

File: packages/chakra-components/src/MenuBar/menubar.js [hunk 11]
```diff
@@ -114,6 +115,9 @@ const MenuBar = forwardRef(
           id={menuBarId}
           role={role}
           ariaLabel={ariaLabel}
+          display="flex"
+          flexDirection={mode && mode === "horizontal" ? "row" : "column"}
+          alignItems={mode && mode === "horizontal" ? "center" : "left"}
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/menubar.js]: Include these prop as style props object like 
```suggestion
let modeStyleProps = {}
if(mode==="vertical) {
   modeStyleProps: {
     flexDirection: column,
     alignItems: left
   } 
}
```

and pass this object after {...styleProps}

File: packages/chakra-components/src/MenuBar/styles.js [hunk 12]
```diff
@@ -6,7 +6,6 @@ import { useColorMode } from "@chakra-ui/core";
 
 export const useMenuBarStyle = () => {
   const baseProps = {
-    alignItems: "center",
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/styles.js]: Please don't change the base styles for a prop change. 
Base styles are touched only if we decide to change the minimal styling.
Include this back

**Review by navin-moorthy — CHANGES_REQUESTED** [review]

File: docs/pages/menubar.mdx [hunk 13]
```diff
@@ -169,3 +169,47 @@ Default trigger - `hover`
...
 ```
+
+### mode and isCollapsable
+
+Default mode - `horizontal`
+
+Default isCollapsable - `false`
+
+```jsx
+<MenuBar mode="vertical" trigger="click">
```
Comment (navin-moorthy) [inline docs/pages/menubar.mdx]:
```suggestion
<MenuBar mode="vertical" trigger="click" maxW={300}>
```

We can include the maxWidth to demonstrate the vertical menubar for example here

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 4 above)
Comment (navin-moorthy) [reply packages/chakra-components/src/MenuBar/SubMenu.js]: These are critical for rendering the focus back to the submenu after blur or mouse leave.
This set of code is necessary.

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 1 above)
Comment (govindsingh55) [reply packages/chakra-components/src/MenuBar/SubMenu.js:5]: rebase with refractor-menubar and it's not being used in this file.

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 2 above)
Comment (govindsingh55) [reply packages/chakra-components/src/MenuBar/SubMenu.js]: rebase with refractor-menubar and it's not being used in this file.

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 4 above)
Comment (govindsingh55) [reply packages/chakra-components/src/MenuBar/SubMenu.js]: rebase with refractor-menubar

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 1 above)
Comment (navin-moorthy) [reply packages/chakra-components/src/MenuBar/SubMenu.js:5]: image

Still used, Bro. 
Please rebase or merge with you branch again.

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 8 above)
Comment (govindsingh55) [reply packages/chakra-components/src/MenuBar/SubMenu.js]: done

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 5 above)
Comment (govindsingh55) [reply packages/chakra-components/src/MenuBar/SubMenu.js]: rebase

File: packages/chakra-components/src/MenuBar/menubar.js (hunk 11 above)
Comment (govindsingh55) [reply packages/chakra-components/src/MenuBar/menubar.js]: done

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 5 above)
Comment (navin-moorthy) [reply packages/chakra-components/src/MenuBar/SubMenu.js]: Rebase or merge now or include those line back bro, they are important.

File: docs/pages/menubar.mdx [hunk 14]
```diff
@@ -144,6 +144,50 @@ import {
...
+
+Default mode - `horizontal`
+
+Default isCollapsable - `false`
+
+```jsx
+<MenuBar mode="vertical" trigger="click" maxWidth="300px">
+  <MenuBarItem>Recipes</MenuBarItem>
+  <MenuBarItem>My Calendar</MenuBarItem>
+  <SubMenu isCollapsable>
```
Comment (navin-moorthy) [inline docs/pages/menubar.mdx]:
```suggestion
  <SubMenu>
```
Just to show one example without collapsable.

File: packages/chakra-components/src/MenuBar/SubMenu.js [hunk 15]
```diff
@@ -190,13 +153,24 @@ const SubMenu = ({
...
+    isCollapsable,
   };
 
+  let modeStyleProps = {};
+  if (isCollapsable) {
+    modeStyleProps = {
+      flexDirection: "column",
+    };
+  }
+
```
Comment (navin-moorthy) [inline packages/chakra-components/src/MenuBar/SubMenu.js]: Looks Great !!

**Comment by navin-moorthy** [comment]
Great work Bro.
Still missing the context and trigger code from [here](https://github.com/timelessco/chakra-components/blob/refractor-menubar/packages/chakra-components/src/MenuBar/SubMenu.js
Kindly add that.

File: packages/chakra-components/src/MenuBar/SubMenuList.js (hunk 10 above)
Comment (navin-moorthy) [reply packages/chakra-components/src/MenuBar/SubMenuList.js]: Make this change and add the create a collapsable style object for placement in Popper.

```js
if(isCollapsable === true) {
  collapsablePopperProp: {
    placement: 'right'
  }
}

this will override the default `placement: 'bottom'`

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 1 above)
Comment (navin-moorthy) [reply packages/chakra-components/src/MenuBar/SubMenu.js:5]: Done 👍

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 2 above)
Comment (navin-moorthy) [reply packages/chakra-components/src/MenuBar/SubMenu.js]: Done 👍

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 3 above)
Comment (navin-moorthy) [reply packages/chakra-components/src/MenuBar/SubMenu.js:48]: Done 👍

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 4 above)
Comment (navin-moorthy) [reply packages/chakra-components/src/MenuBar/SubMenu.js]: Done 👍

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 5 above)
Comment (navin-moorthy) [reply packages/chakra-components/src/MenuBar/SubMenu.js]: Done 👍

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 6 above)
Comment (navin-moorthy) [reply packages/chakra-components/src/MenuBar/SubMenu.js]: Seems this is not yer added back.
Please add it bro.

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 7 above)
Comment (navin-moorthy) [reply packages/chakra-components/src/MenuBar/SubMenu.js:181]: Kindly add this as well. 
Go through and understand, what this function does and what doesn't happen removing this.
Let me know if you don't understand.

File: packages/chakra-components/src/MenuBar/SubMenu.js (hunk 6 above)
Comment (govindsingh55) [reply packages/chakra-components/src/MenuBar/SubMenu.js]: added

File: docs/pages/menubar.mdx (hunk 14 above)
Comment (navin-moorthy) [reply docs/pages/menubar.mdx]: Done

**Review by navin-moorthy — APPROVED** [review]
